File: jzoj/1228.pas
const	maxn=1000+5;
var	i,j,m,n,cnt,x,ans:longint;
	a:array[1..maxn,1..maxn] of shortint;
	b,c:array[1..maxn] of shortint;		//ÊÇ·ñ½»»»
function check:boolean;
var	i,j,tmp:longint;
begin
	for i:=1 to n do
		for j:=1 to m do begin
			tmp:=c[i] xor b[j];
			if tmp<>a[i,j] then exit(false);		
		end;
	exit(true);
end;
begin
	assign(input,'1228.in');reset(input);
	//assign(output,'.out');rewrite(output);

	readln(n,m);
	for i:=1 to n do
		for j:=1 to m do begin
			read(x);
			if x>0 then a[i,j]:=0 else a[i,j]:=1;
		end;
	
	cnt:=0;ans:=-1;
	for j:=1 to m do
	       	if a[1][j]=1 then b[j]:=0 else begin
			b[j]:=1;inc(cnt);
		end;
	for j:=1 to n do
		if b[1]=1 then begin
			if a[j,1]=1 then c[j]:=0 else begin
				c[j]:=1;
				inc(cnt);
			end;
		end else begin
			if a[j][1]=1 then begin
				c[j]:=1;
				inc(cnt);
			end else c[j]:=0;	
		end;
	if ((ans=-1) or (cnt<=ans)) and (check) then ans:=cnt;

	cnt:=0;
	for j:=1 to m do
	       	if a[1][j]=0 then b[j]:=0 else begin
			b[j]:=1;inc(cnt);
		end;
	for j:=1 to n do
		if b[1]=1 then begin
			if a[j,1]=1 then c[j]:=0 else begin
				c[j]:=1;
				inc(cnt);
			end;
		end else begin
			if a[j][1]=1 then begin
				c[j]:=1;
				inc(cnt);
			end else c[j]:=0;	
		end;
	if ((ans=-1) or (cnt<=ans)) and (check) then ans:=cnt;
	
	if ans=-1 then writeln('impossible') else writeln(ans);
end.

File: OI/jzoj/1228.cpp
#include<cstring>
#include<cstdio>
#include<algorithm>
using namespace std;

int a[1005][1005],b[1005],c[1005],n,m,ans;

int Check()
{
	for (int i = 1 ; i <= n ; i ++)
		for (int j = 1 ; j <= m ; j ++)
		{
			int tmp = (c[i] ^ b[j]);
			if (tmp != a[i][j]) return 0;
		}
	return 1;
}

int main()
{
	freopen("1228.in","r",stdin);
	scanf("%d%d" , &n , &m);
	for (int i = 1 ; i <= n ; i ++)
		for (int j = 1 ; j <= m ; j ++)
		{
			int x ; scanf("%d" , &x);
			if (x > 0) a[i][j] = 0 ; else a[i][j] = 1;
		}
	
	ans = -1;
	for (int i = 1 ; i <= 1 ; i ++)
	{
		int tp = 0; // if i % 2 == 1
		for (int j = 1 ; j <= m ; j ++) if (a[i][j]) b[j] = 0 ; else 
			b[j] = 1 , tp ++;
		for (int j = 1 ; j <= n ; j ++) if (b[1])
		{
			if (a[j][1]) c[j] = 0 ; else c[j] = 1 , tp ++;
		} else
		{
			if (a[j][1]) c[j] = 1 , tp ++ ; else c[j] = 0;
		}
		//if (c[i] != 1) continue;
		if (ans != -1 && tp >= ans) continue;
		if (!Check()) continue;
		ans = tp;
	}
	
	for (int i = 1 ; i <= 1 ; i ++)
	{
		int tp = 0; // if i % 2 == 1
		for (int j = 1 ; j <= m ; j ++) if (a[i][j]) b[j] = 1 , tp ++; else 
			b[j] = 0;
		for (int j = 1 ; j <= n ; j ++) if (b[1])
		{
			if (a[j][1]) c[j] = 0 ; else c[j] = 1 , tp ++;
		} else
		{
			if (a[j][1]) c[j] = 1 , tp ++ ; else c[j] = 0;
		}
		//if (c[i] != 0) continue;
		if (ans != -1 && tp >= ans) continue;
		if (!Check()) continue;
		ans = tp;
	}
	
	if (ans == -1) printf("impossible") ; else printf("%d" , ans);
	
	return 0;
}
